src/client/particles.h

```diff
@@ -164,20 +149,62 @@ class ParticleSpawner
 	size_t m_active;
 	ParticleManager *m_particlemanager;
 	float m_time;
-	IGameDef *m_gamedef;
 	LocalPlayer *m_player;
 	ParticleSpawnerParameters p;
 	std::vector<ClientParticleTexture> m_texpool;
 	std::vector<float> m_spawntimes;
 	u16 m_attached_id;
 };
 
+class ParticleBuffer : public scene::ISceneNode
+{
+	friend class ParticleManager;
+public:
+	ParticleBuffer(ClientEnvironment *env, const video::SMaterial &material);
+
+	// for pointer stability
+	DISABLE_CLASS_COPY(ParticleBuffer)
+
+	std::optional<u16> allocate();
+	void release(u16 index);
+
+	// video::S3DVertex[4]
+	video::S3DVertex *getVertices(u16 index);
+
+	inline bool isEmpty() const {
+		return m_free_list.size() == m_count;
+	}
+
+	virtual video::SMaterial &getMaterial(u32 num) override {
+		return m_mesh_buffer->getMaterial();
+	}
+	virtual u32 getMaterialCount() const override {
+		return 1;
+	}
+
+	virtual const core::aabbox3df &getBoundingBox() const override;
+
+	virtual void render() override;
+
+	virtual void OnRegisterSceneNode() override;
+
+	// we have 16-bit indices
+	static constexpr u16 MAX_PARTICLES_PER_BUFFER = 16000;
+
+private:
+	irr_ptr<scene::SMeshBuffer> m_mesh_buffer;
+	std::vector<u16> m_free_list;
+	float m_usage_timer = 0; // reset on allocation
+	u16 m_count = 0;
+	mutable bool m_bounding_box_dirty = true;
```

Comment:
```suggestion
	std::vector<u16> m_free_list; //< unused (e.g. expired) particle indices for re-use
	float m_usage_timer = 0; //< for automatic deletion when unused for a while
	u16 m_count = 0; //< total count of contained particles
	mutable bool m_bounding_box_dirty = true;
```

Why does `m_count` exist? `m_mesh_buffer->getVertexCount() / 4 - m_free_list.size()` would never run out of sync.

Reply:
I think keeping track of this is ultimately simpler than inferring it from other state.

Reply:
I just noticed that `m_count` is in fact the overall capacity (ignoring the free ones), thus `m_mesh_buffer->getVertexCount() / 4`. Divisions by 4 should be very quick to calculate. However, I agree that the current way is simpler and likely faster. I'll leave the decision up to you. Some documentation as suggested above would be very helpful, though.